sqlanalytics/resource_query.go

```diff
@@ -0,0 +1,495 @@
+package sqlanalytics
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"reflect"
+	"strings"
+
+	"github.com/databrickslabs/terraform-provider-databricks/common"
+	"github.com/databrickslabs/terraform-provider-databricks/sqlanalytics/api"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// QueryEntity defines the parameters that can be set in the resource.
+type QueryEntity struct {
+	DataSourceID string           `json:"data_source_id"`
+	Name         string           `json:"name"`
+	Description  string           `json:"description,omitempty"`
+	Query        string           `json:"query"`
+	Schedule     *QuerySchedule   `json:"schedule,omitempty"`
+	Tags         []string         `json:"tags,omitempty"`
+	Parameter    []QueryParameter `json:"parameter,omitempty"`
+}
+
+// QuerySchedule ...
+type QuerySchedule struct {
+	Interval int `json:"interval"`
+}
+
+// QueryParameter ...
+type QueryParameter struct {
+	Name  string `json:"name"`
+	Title string `json:"title,omitempty"`
+
+	// Type specific structs.
+	// Only one of them may be set.
+	Text             *QueryParameterText             `json:"text,omitempty"`
+	Number           *QueryParameterNumber           `json:"number,omitempty"`
+	Enum             *QueryParameterEnum             `json:"enum,omitempty"`
+	Query            *QueryParameterQuery            `json:"query,omitempty"`
+	Date             *QueryParameterDate             `json:"date,omitempty"`
+	DateTime         *QueryParameterDateTime         `json:"datetime,omitempty"`
+	DateTimeSec      *QueryParameterDateTimeSec      `json:"datetimesec,omitempty"`
+	DateRange        *QueryParameterDateRange        `json:"date_range,omitempty"`
+	DateTimeRange    *QueryParameterDateTimeRange    `json:"datetime_range,omitempty"`
+	DateTimeSecRange *QueryParameterDateTimeSecRange `json:"datetimesec_range,omitempty"`
+}
+
+// QueryParameterText ...
+type QueryParameterText struct {
+	Value string `json:"value"`
+}
+
+// QueryParameterNumber ...
+type QueryParameterNumber struct {
+	Value float64 `json:"value"`
+}
+
+// QueryParameterEnum ...
+type QueryParameterEnum struct {
+	// Value iff `multiple == nil`
+	Value string `json:"value,omitempty"`
+	// Values iff `multiple != nil`
+	Values []string `json:"values,omitempty"`
+
+	Options  []string                     `json:"options"`
+	Multiple *QueryParameterAllowMultiple `json:"multiple,omitempty"`
+}
+
+// QueryParameterQuery ...
+type QueryParameterQuery struct {
+	// Value iff `multiple == nil`
+	Value string `json:"value,omitempty"`
+	// Values iff `multiple != nil`
+	Values []string `json:"values,omitempty"`
+
+	QueryID  string                       `json:"query_id"`
+	Multiple *QueryParameterAllowMultiple `json:"multiple,omitempty"`
+}
+
+// QueryParameterDate ...
+type QueryParameterDate struct {
```

Comment:
why do we need so many types with literally the same structure?...

Reply:
I started with a single struct per query parameter type.

Checked it out and was able to collapse the non-range and range date-like types into a single struct.